python/triton/compiler/compiler.py

```diff
@@ -379,14 +300,14 @@ def compile(fn, **kwargs):
     stages["ast"] = (lambda path: fn, None)
     stages["ttir"] = (lambda path: parse_mlir_module(path, context),
                       lambda src: optimize_ttir(ast_to_ttir(src, signature, configs[0], constants, debug=debug, arch=arch), arch))
-    stages["ttgir"] = (lambda path: parse_mlir_module(path, context),
-                       lambda src: optimize_ttgir(ttir_to_ttgir(src, num_warps), num_stages, arch))
-    stages["llir"] = (lambda path: Path(path).read_text(),
-                      lambda src: ttgir_to_llir(src, extern_libs, arch))
     if is_cuda:
-        add_cuda_stages(arch, extern_libs, stages)
-    elif is_hip:
-        add_rocm_stages(arch, extern_libs, stages)
+        add_cuda_stages(arch, context, num_warps, num_stages, extern_libs, stages)
+    elif is_hip():
+        _device_backend.add_stages(arch, extern_libs, stages, context=context, num_warps=num_warps, num_stages=num_stages)
+    elif device_type == "xpu":
+        stages["ttgir"] = (lambda path: parse_mlir_module(path, context),
+                           lambda src: optimize_ttgir(ttir_to_ttgir(src, num_warps), num_stages, arch))
+        _device_backend.add_stages(arch, extern_libs, stages)
```

Comment:
why are `xpu` changes necessary in a patch for AMD?

Reply:
The common stages for all the backends are ast and ttir. we use our own gpu dialect and xpu uses the ttgir generated for the cuda backend. that is why I moved the code there. From an xpu prespective everything should be the same.